### DISCUSSION TITLE
How do I set up an ltsp-server in an lxd container

### DISCUSSION THREAD

**Discussion by emsfree** [ltsp/ltsp#581] [Q&A] [answered]
Hey there,
I am not quite sure whether this is the right place... but I would like to set up an ltsp-server in an lxd- container and find my self stuck with the basics...
Has anyone done this or could link me to usable sources?
The host and LTSP would preferably be ubuntu 20.04.
Thanks so much for any help!
emsfree

**Comment by alkisg** [comment, marked as answer]
There are two ways to approach this issue...
One is to try to find someone that knows both ltsp and lxd, and is able to help you. Personally I haven't tried lxd at all.
And the other way is to learn about ltsp and lxd separately, and then combine the knowledge yourself.
For ltsp, the best way to start is https://ltsp.org/docs/installation/, if you already have e.g. an Ubuntu installation it then takes 20 minutes to do your first ltsp installation and learn the ltsp basics.

**Reply by emsfree** [reply]
Thank you very much! Since I have already some experience with LTSP (thank you for your magnificent work!) I will then dive into LXD Containers and take it from scratch. I would like to create a template or image for a container with ltsp for others to use it as well. If anyone is interested or has tips I would be very grateful!

**Reply by jmgibson1981** [reply]
I had this running some time ago. I did try the lxc direct options and i wasn't happy with it. I'm not sure if you've already solved or moved on. Here are the lines I had to run on the LXD container to make it work.
 I never had success with the config set raw.* values.
```
lxc config set "$1" security.privileged true
lxc config set "$1" security.nesting true
```

**Comment by ecfsolutions** [comment]
Hello emsfree
there you go like a bull at a gate with me ... I tried this these days and failed because as far as I understood LXD/LXC has problems with nfs-shares needed for pxe boot - if you are interested, we can gladly exchange our experiences / throw our knowledge together ...

**Reply by emsfree** [reply]
Did you get my mail? Sorry, I though it would also appear here.
Currently, I am busy setting up a first try with the settings know-how I collected so far. It's more a guess, though.
My idear is to set up a LTSP in a container, but provide the client-image (via symlink) as virtual machine (suggested by LTSP [https://ltsp.org/man/ltsp-image/]. I understand that NFS probably needs to run outside the container and will be back here as soon as I learned more to ask specific questions. Thanks!

**Comment by Mogaba** [comment]
If it's still relevant: I run LTSP in a privileged LXC container without any issues. I don't use LXD so I don't know if there're any pitfalls. What I do know: you will have a lot of headache if you try to run LTSP in an unprivileged container.

**Reply by vsuojanen** [reply]
Here also running LTSP servers in privileged LXC containers works fine. NFS or any PXE boot parameters should not have any issues. 

'lxc config set NAME security.privileged true'

First the 'ltsp image' could fail when some container images have Apparmor protect rules:
'lxc config set NAME raw.apparmor "mount,"' 

Second if 'ltsp image' fails - make sure  to provide loop devices in container:
lxc config device add NAME loop0 unix-block path=/dev/loop0
lxc config device add NAME loop1 unix-block path=/dev/loop1
...
lxc config device add NAME loop-control unix-char path=/dev/loop-control

Third if 'ltsp image' fails to mounting /boot from the created image (at the very end when updating kernel and initrd from the image to tftp) install these to the container (if chrootless image setup) or if using other image source to the chroot environment:
apt install --no-install-recommends linux-generic initramfs-tools

**Reply by emsfree** [reply]
Thank you all for the answers! This will be the basis in a few weeks.
I put the project on hold for the moment and used a dedicated install of ubuntu as ltsp-server (the usual setup) instead. In the next round, I will focus on the containers again!
Thanks again for the contributions!
